fdbserver/MoveKeys.actor.cpp

```diff
@@ -1306,11 +1351,68 @@ ACTOR Future<Void> removeKeysFromFailedServer(Database cx,
 					src.erase(std::remove(src.begin(), src.end(), serverID), src.end());
 					dest.erase(std::remove(dest.begin(), dest.end(), serverID), dest.end());
-					TraceEvent(SevDebug, "FailedServerSetKey", serverID)
-					    .detail("Key", it.key)
-					    .detail("ValueSrc", describe(src))
-					    .detail("ValueDest", describe(dest));
-					tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, src, dest));
+
+					// If the last src server is to be removed, first check if there are dest servers who is
+					// hosting a read-write copy of the keyrange, and move such dest servers to the src list.
+					if (src.empty() && !dest.empty()) {
+						std::vector<UID> newSources =
+						    wait(pickReadWriteServers(&tr, dest, KeyRangeRef(it.key, keyServers[i + 1].key)));
+						for (const UID& id : newSources) {
+							TraceEvent(SevWarn, "FailedServerAdditionalSourceServer", serverID)
+							    .detail("Key", it.key)
+							    .detail("NewSourceServerFromDest", id);
+							dest.erase(std::remove(dest.begin(), dest.end(), id), dest.end());
+							src.push_back(id);
+						}
+					}
+
+					// Move the keyrange to teamForDroppedRange if the src list becomes empty, and also remove the shard
+					// from all dest servers.
+					if (src.empty()) {
+						assert(!teamForDroppedRange.empty());
+
+						// Assign the shard to teamFroDroppedRange in keyServer space.
+						tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, teamForDroppedRange, {}));
+
+						vector<Future<Void>> actors;
+
+						// Unassign the shard from the dest servers.
+						for (const UID& id : dest) {
+							actors.push_back(krmSetRangeCoalescing(&tr,
+							                                       serverKeysPrefixFor(id),
+							                                       KeyRangeRef(it.key, keyServers[i + 1].key),
+							                                       allKeys,
+							                                       serverKeysFalse));
+						}
+						if (!dest.empty()) {
+							TraceEvent(SevWarn, "FailedServerDropRangeFromDest", serverID)
+							    .detail("Begin", it.key)
+							    .detail("End", keyServers[i + 1].key)
+							    .detail("Dest", describe(dest));
+						}
+
+						// Assign the shard to the new team as an empty range.
+						// Note, there could be data loss.
+						for (const UID& id : teamForDroppedRange) {
+							actors.push_back(krmSetRangeCoalescing(&tr,
+							                                       serverKeysPrefixFor(id),
+							                                       KeyRangeRef(it.key, keyServers[i + 1].key),
+							                                       allKeys,
+							                                       serverKeysTrueEmptyRange));
+						}
+
+						wait(waitForAll(actors));
+						TraceEvent(SevWarn, "FailedServerDropRange", serverID)
+						    .detail("Begin", it.key)
+						    .detail("End", keyServers[i + 1].key)
+						    .detail("NewTeam", describe(teamForDroppedRange));
```

Comment:
Can we add a field on
which destination SSes have been dropped for the shard?

Reply:
Combined the above trace with this one.